### PR TITLE
[4.1] Add Visual Indicator for Failing Scheduled Tasks

### PR DESCRIPTION
**Pull request by ditsuke**
Pull Request for Issue #36459.

### Summary of Changes
- Adds an icon as a visual indicator for failing tasks.
- Includes an on-hover tooltip for accessibility (review).

### Testing Instructions
- Create a new scheduled task with routine GET Request, setting an invalid URL as the target
.
- Let it run. It will fail.

### Actual result BEFORE applying this Pull Request
There is no visual indicator that the task has failed in the default Scheduled tasks view.

before-patch

### Expected result AFTER applying this Pull Request
There is a visual indicaator that the task is failing:
after-patch

### Documentation Changes Required

None

### CODE REVIEW

**Comment by brianteeman** [comment]
Please look at other "tooltips" and create this one in the same way.

```
	<span class="icon-info-circle" aria-hidden="true" tabindex="0"></span>
	<div role="tooltip" class="text-start" id="tip<?php echo $language->code; ?>">
		<?php echo Text::_('JGLOBAL_LANGUAGE_VERSION_NOT_PLATFORM'); ?>
	</div>

```

**Comment by richard67** [comment]
@ditsuke Could you check the [previous comment](https://github.com/joomla/joomla-cms/pull/36765#issuecomment-1018281670)? Thanks in advance.

File: administrator/components/com_scheduler/tmpl/tasks/default.php [hunk 1]
```diff
@@ -198,12 +202,20 @@ class="js-draggable" data-url="<?php echo $saveOrderingUrl; ?>" data-direction="
...
-							<?php if ($canEdit): ?>
-								<a href="<?php echo Route::_('index.php?option=com_scheduler&task=task.edit&id=' . $item->id); ?>"
-								   title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape($item->title); ?>"> <?php echo $this->escape($item->title); ?></a>
-							<?php else: ?>
-								<?php echo $this->escape($item->title); ?>
-							<?php endif; ?>
+							<span class="task-title">
+								<?php if ($canEdit): ?>
+									 <a href="<?php echo Route::_('index.php?option=com_scheduler&task=task.edit&id=' . $item->id); ?>"
+										 title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape($item->title); ?>"> <?php echo $this->escape($item->title); ?></a>
```
Comment (richard67) [inline administrator/components/com_scheduler/tmpl/tasks/default.php]:
```suggestion
									<a href="<?php echo Route::_('index.php?option=com_scheduler&task=task.edit&id=' . $item->id); ?>"
										title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape($item->title); ?>"> <?php echo $this->escape($item->title); ?>
									</a>
```

File: administrator/components/com_scheduler/tmpl/tasks/default.php [hunk 2]
```diff
@@ -198,12 +202,20 @@ class="js-draggable" data-url="<?php echo $saveOrderingUrl; ?>" data-direction="
...
+								<?php else: ?>
+									 <?php echo $this->escape($item->title); ?>
+								<?php endif; ?>
+								<?php if (!in_array($item->last_exit_code, [Status::OK, Status::WILL_RESUME])): ?>
+									<span class="failure-indicator icon-solid icon-exclamation-triangle hasTooltip" aria-hidden="true"></span>
+									<div role="tooltip">
+        								<?php echo Text::sprintf("COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING", $item->last_exit_code); ?>
+    								</div>
+								<?php endif; ?>
```
Comment (richard67) [inline administrator/components/com_scheduler/tmpl/tasks/default.php]:
```suggestion
								<?php if (!in_array($item->last_exit_code, [Status::OK, Status::WILL_RESUME])): ?>
									<span class="failure-indicator icon-solid icon-exclamation-triangle hasTooltip" aria-hidden="true"></span>
									<div role="tooltip">
										<?php echo Text::sprintf("COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING", $item->last_exit_code); ?>
									</div>
								<?php endif; ?>
```

File: administrator/components/com_scheduler/tmpl/tasks/default.php (hunk 1 above)
Comment (richard67) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: Remove space before the anchor and use tabs and not spaced to indent.

File: administrator/components/com_scheduler/tmpl/tasks/default.php (hunk 2 above)
Comment (richard67) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: Use consistent indent.

Comment (richard67) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: This will also fix drone, I think.

Comment (richard67) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: There were also spaces and not tabs.

Comment (ditsuke) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: Thanks

File: administrator/components/com_scheduler/tmpl/tasks/default.php (hunk 1 above)
Comment (richard67) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: @ditsuke Drone PHPCS doesn't complain about this one, but it should be fixed, too.

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 3febb572a01b0f52e381c148de4a7d3afc212f40<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

File: administrator/components/com_scheduler/tmpl/tasks/default.php [hunk 3]
```diff
@@ -62,7 +63,10 @@
 
 $document = $app->getDocument();
 $document->addScriptOptions('com_scheduler.test-task.token', Session::getFormToken());
-$document->getWebAssetManager()->useScript('com_scheduler.test-task');
+
+$wa = $document->getWebAssetManager();
+$wa->useScript('com_scheduler.test-task');
+$wa->useStyle('com_scheduler.admin-view-tasks-css');
```
Comment (heelc29) [inline administrator/components/com_scheduler/tmpl/tasks/default.php]:
```suggestion
/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
$wa = $this->document->getWebAssetManager();
$wa->useScript('com_scheduler.test-task')
	->useStyle('com_scheduler.admin-view-tasks-css');
```

File: administrator/components/com_scheduler/tmpl/tasks/default.php [hunk 4]
```diff
@@ -62,7 +63,10 @@
 
 $document = $app->getDocument();
 $document->addScriptOptions('com_scheduler.test-task.token', Session::getFormToken());
```
Comment (heelc29) [inline administrator/components/com_scheduler/tmpl/tasks/default.php]:
```suggestion
$this->document->addScriptOptions('com_scheduler.test-task.token', Session::getFormToken());
```

**Review by heelc29 — APPROVED** [review]

File: administrator/components/com_scheduler/tmpl/tasks/default.php (hunk 4 above)
Comment (ditsuke) [reply administrator/components/com_scheduler/tmpl/tasks/default.php]: @heelc29 thanks!

**Comment by pritam825** [comment]
Screenshot (1981)
After applying the patch it' giving me error with this -> "There is no "com_scheduler.admin-view-tasks-css" asset of a "style" type in the registry".

**Comment by pritam825** [comment]
I think you forgot to declare this -> "com_scheduler.admin-view-tasks-css"

**Comment by richard67** [comment]
@pritam825 Have you "compiled" the css with `npm ci`or `npm run build:css`as described here https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment after having applied the patch? This will copy the css from "build/media_source/com_scheduler/css" to "media/com_scheduler/css".

**Comment by pritam825** [comment]
> @pritam825 Have you "compiled" the css with `npm ci`or `npm run build:css`as described here https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment after having applied the patch? This will copy the css from "build/media_source/com_scheduler/css" to "media/com_scheduler/css".

sorry sir, I forgot to use npm run build:css, Thanks sir

**Comment by pritam825** [comment]
It's still giving me the same error

**Comment by richard67** [comment]
> It's still giving me the same error

@pritam825 Try `npm ci` instead of `npm run build:css` and clear browser cache before testing again.

**Comment by pritam825** [comment]
@richard67  thanks, now it's working fine.
Working as expected
Screenshot (1984)

**Comment by richard67** [comment]
@pritam825 Then mark your test result in the issue tracker. Thanks in advance.

**Comment by pritam825** [comment]
ok sir, Thanks

**Comment by pritam825** [comment]
I have tested this item :white_check_mark: successfully on a9d6300100f69f67b5cf9d0fc9cd1157c0bafe55<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

**Comment by richard67** [comment]
@chmst Could you test again? Thanks in advance.

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on a9d6300100f69f67b5cf9d0fc9cd1157c0bafe55<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

**Comment by jwaisner** [comment]
Back to pending.

@ditsuke ,

 Can you please fix the conflicts?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

**Comment by richard67** [comment]
@ditsuke @jwaisner I've allowed myself to fix the merge conflict and to add back the previous results since the change after those just were a clean branch update with the conflict fix. => RTC again.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

**Comment by ditsuke** [comment]
@richard67 @jwaisner thanks!

File: administrator/components/com_scheduler/tmpl/tasks/default.php [hunk 5]
```diff
@@ -199,12 +201,21 @@ class="js-draggable" data-url="<?php echo $saveOrderingUrl; ?>" data-direction="
...
+							<span class="task-title">
+								<?php if ($canEdit): ?>
+									<a href="<?php echo Route::_('index.php?option=com_scheduler&task=task.edit&id=' . $item->id); ?>"
+										title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape($item->title); ?>"> <?php echo $this->escape($item->title); ?>
+									</a>
+								<?php else: ?>
+									 <?php echo $this->escape($item->title); ?>
+								<?php endif; ?>
+								<?php if (!in_array($item->last_exit_code, [Status::OK, Status::WILL_RESUME])): ?>
+									<span class="failure-indicator icon-solid icon-exclamation-triangle hasTooltip" aria-hidden="true"></span>
```
Comment (Quy) [inline administrator/components/com_scheduler/tmpl/tasks/default.php]:
```suggestion
									<span class="failure-indicator icon-exclamation-triangle" aria-hidden="true"></span>
```

File: administrator/language/en-GB/com_scheduler.ini [hunk 6]
```diff
@@ -81,6 +81,7 @@ COM_SCHEDULER_LAST_RUN_DATE="Last Run Date"
 COM_SCHEDULER_MANAGER_TASKS="Scheduled Tasks"
 COM_SCHEDULER_MANAGER_TASK_EDIT="Edit Task"
 COM_SCHEDULER_MANAGER_TASK_NEW="New Task"
+COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING="Task is failing with exit code %1$d"
```
Comment (Quy) [inline administrator/language/en-GB/com_scheduler.ini]:
```suggestion
COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING="Task failed. Exit code: %1$d"
```

**Comment by richard67** [comment]
@ditsuke Could you implement @Quy 's suggestions? Thanks in advance.

**Comment by ditsuke** [comment]
@richard67 applied

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 05dde691b766a0c2521e992b10ef638ddb5e3208<br /><br />I've restored the previous test results since the change after that was just removal of 2 unused CSS classes in the markup and an improvement on the language string for the tool tip.

In addition I have tested it, too, just to be sure.

=> RTC still valid.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

**Comment by bembelimen** [comment]
Thx

**Comment by ditsuke** [comment]
Thanks!

**Pull request merged**